spec/ruby/core/numeric/shared/step.rb

```diff
@@ -0,0 +1,397 @@
+require File.expand_path('../../../../spec_helper', __FILE__)
+require File.expand_path('../../helpers/step', __FILE__)
+require File.expand_path('../../fixtures/classes', __FILE__)
+
+describe :numeric_step, :shared => true do
+  before :each do
+    ScratchPad.record []
+    @prc = lambda { |x| ScratchPad << x }
+  end
+
+  it "defaults to step = 1" do
+    1.send(@method, *get_args(@args_type, 5), &@prc)
+    ScratchPad.recorded.should == [1, 2, 3, 4, 5]
+  end
+
+  it "returns an Enumerator when step is 0" do
+    1.send(@method, *get_args(@args_type, 2, 0)).should be_an_instance_of(enumerator_class)
+  end
+
+  it "returns an Enumerator when not passed a block and self > stop" do
+    1.send(@method, *get_args(@args_type, 0, 2)).should be_an_instance_of(enumerator_class)
+  end
+
+  it "returns an Enumerator when not passed a block and self < stop" do
+    1.send(@method, *get_args(@args_type, 2, 3)).should be_an_instance_of(enumerator_class)
+  end
+
+  it "returns an Enumerator that uses the given step" do
+    0.send(@method, *get_args(@args_type, 5, 2)).to_a.should == [0, 2, 4]
+  end
+
+  describe "with [stop, step]" do
+    before :each do
+      @stop = mock("Numeric#step stop value")
+      @step = mock("Numeric#step step value")
+      @obj = NumericSpecs::Subclass.new
+    end
+
+    it "increments self using #+ until self > stop when step > 0" do
+      @step.should_receive(:>).with(0).and_return(true)
+      @obj.should_receive(:>).with(@stop).and_return(false, false, false, true)
+      @obj.should_receive(:+).with(@step).and_return(@obj, @obj, @obj)
+
+      @obj.send(@method, *get_args(@args_type, @stop, @step), &@prc)
+
+      ScratchPad.recorded.should == [@obj, @obj, @obj]
```

Comment:
These tests don't test behaviour, they test implementation details. Specs shouldn't dictate what methods internally are used except for a few rare edge cases where calling a certain method is part of the defined behaviour. In this case it's crazy to test that the objects called `>` and `+`.